docs/SPEC.md

```diff
@@ -1370,6 +1371,115 @@ A query specification defines what data and operations to perform.
 The execution model reserves the right to perform those operations as efficiently as possible.
 The execution model may rewrite the query in anyway it sees fit while maintaining correctness.
 
+## Modules
+
+>NOTE: Modules are not fully implemented yet, follow https://github.com/influxdata/flux/issues/4296 for details.
+
+A module is a collection of packages that can be imported.
+A module has a module path, version and a collection of packages with their source code.
+
+### Module path
+
+The module path is the import path of the top level package within the module.
+Additionally major versions of a module of two or greater must add a final element to the path of the form `v#` where `#` is the major version number.
+For modules at version zero or one the path must not contain the major version as it is not necessary.
+A change from `v0` to `v1` may include a breaking change but once `v1` is published any future breaking changes will be a new major version.
+
+Example
+
+    foo/bar    // module path of foo/bar for version zero or one
+    foo/bar/v2 // module path of foo/bar for major version two
+
+### Module versions
+
+All modules are versioned using a [semantic version number](https://semver.org/) prefixed with a `v`, i.e. `vMAJOR.MINOR.PATCH`.
+Once a module version has been published it cannot be modified.
+A new version of the module must be published.
+
+### Importing modules
```

Comment:
I wonder about how users would elect to load from local disk. Would it be some prefix or edit in the import path (which would be a tedious workflow) or would we have a way to set custom "resolve" (as with npm) or "registry patches" (as with cargo)?

Reply:
I think with the new specification this would look like

```
@registry("modules", "file://./fluxmods")
package main

import "modules/foo"
```

Now the URL for `foo` will be `file://./fluxmods/foo` and we can implement the logic to grab it off disk. The granularity here might be too coarse since its the entire registry but a `registry_patch` or similar attribute could be used in those cases. In short the fundamentals is we map registry names to URLs and those URLs can point to the local filesystem.
```

Reply:
Yeah, I think this works for me.

Though I have some misgivings about the file protocol appearing anywhere in the flux source code since _I think_ we'd need to start thinking about security implications.

Less concerned about this if there is a patching mechanism that has some guardrails around what paths are allowed for reading (before we even look at the source code).